File: BackEnd/src/main/java/com/talentpredict/modules/assessment/services/CampaignService.java
package com.talentpredict.modules.assessment.services;

import java.util.List;
import java.util.UUID;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import com.talentpredict.modules.assessment.dto.CampaignDto;
import com.talentpredict.modules.assessment.entities.Campaign;
import com.talentpredict.modules.assessment.repositories.CampaignRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CampaignService {

    private final CampaignRepository campaignRepository;

    @Transactional(readOnly = true)
    public List<CampaignDto> listCampaigns() {
        return campaignRepository.findAllByOrderByCreatedAtDesc().stream()
                .map(this::toDto)
                .toList();
    }

    @Transactional(readOnly = true)
    public CampaignDto findById(UUID id) {
        return campaignRepository.findById(id)
                .map(this::toDto)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Campaign not found"));
    }

    @Transactional
    public CampaignDto saveCampaign(CampaignDto request) {
        Campaign entity = request.id() != null
                ? campaignRepository.findById(request.id()).orElseGet(Campaign::new)
                : new Campaign();

        if (entity.getId() == null && request.id() != null) {
            entity.setId(request.id());
        }

        apply(entity, request);
        Campaign saved = campaignRepository.save(entity);
        return toDto(saved);
    }

    @Transactional
    public CampaignDto update(UUID id, CampaignDto request) {
        Campaign entity = campaignRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Campaign not found"));
        
        apply(entity, request);
        Campaign saved = campaignRepository.save(entity);
        return toDto(saved);
    }

    @Transactional
    public void delete(UUID id) {
        if (!campaignRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Campaign not found");
        }
        campaignRepository.deleteById(id);
    }

    private void apply(Campaign entity, CampaignDto request) {
        entity.setName(trimOrNull(request.name()));
        entity.setTemplateId(trimOrNull(request.templateId()));
        entity.setTemplateName(trimOrNull(request.templateName()));
        entity.setChannel(trimOrNull(request.channel()));
        entity.setTargetGroup(trimOrNull(request.targetGroup()));
        entity.setStatus(trimOrNull(request.status()));
        entity.setScheduledAt(request.scheduledAt());
        entity.setRecipientCount(defaultInt(request.recipientCount()));
        entity.setSentCount(defaultInt(request.sentCount()));
        entity.setFailedCount(defaultInt(request.failedCount()));
        entity.setOpenRate(request.openRate());
        entity.setClickRate(request.clickRate());
        entity.setPaused(defaultBool(request.isPaused()));
    }

    private CampaignDto toDto(Campaign entity) {
        return new CampaignDto(
                entity.getId(),
                entity.getName(),
                entity.getTemplateId(),
                entity.getTemplateName(),
                entity.getChannel(),
                entity.getTargetGroup(),
                entity.getRecipientCount(),
                entity.getStatus(),
                entity.getScheduledAt(),
                entity.getSentCount(),
                entity.getFailedCount(),
                entity.getOpenRate(),
                entity.getClickRate(),
                entity.getPaused());
    }

    private String trimOrNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private Integer defaultInt(Integer value) {
        return value == null ? 0 : value;
    }

    private Boolean defaultBool(Boolean value) {
        return Boolean.TRUE.equals(value);
    }
}

File: FrontEnd/src/app/modules/admin/services/campaign.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';

export interface CampaignApi {
  id: string;
  name: string;
  templateId: string;
  templateName: string;
  channel: string;
  targetGroup: string;
  recipientCount: number;
  status: string;
  scheduledAt: string | null;
  sentCount: number;
  failedCount: number;
  openRate?: number;
  clickRate?: number;
  isPaused?: boolean;
}

export type CampaignUpsertRequest = Omit<CampaignApi, 'id'> & { id?: string };

@Injectable({ providedIn: 'root' })
export class CampaignService {
  private http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/admin/campaigns`;

  listCampaigns(): Observable<CampaignApi[]> {
    return this.http.get<CampaignApi[]>(this.baseUrl);
  }

  getById(id: string): Observable<CampaignApi> {
    return this.http.get<CampaignApi>(`${this.baseUrl}/${id}`);
  }

  saveCampaign(payload: CampaignUpsertRequest): Observable<CampaignApi> {
    return this.http.post<CampaignApi>(this.baseUrl, payload);
  }

  update(id: string, data: CampaignUpsertRequest): Observable<CampaignApi> {
    return this.http.put<CampaignApi>(`${this.baseUrl}/${id}`, data);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
